File: scripts/analyze_transactions.py
import requests
import csv
from io import StringIO
from datetime import datetime

# CSVファイルのURLを取得
csv_url = "https://hebbkx1anhila5yf.public.blob.vercel-storage.com/export-token-transfer-0x77247CC270768611eb2fBc7759a7b34b9FB045Cd-aNDOvRIizWcT7P9hYKPFvRswZi1dDS.csv"

print("[v0] Fetching transaction history CSV...")
response = requests.get(csv_url)
response.raise_for_status()

# CSVデータを解析
csv_data = StringIO(response.text)
reader = csv.DictReader(csv_data)

print("\n[v0] Transaction History Analysis")
print("=" * 80)

transactions = []
for row in reader:
    transactions.append(row)

print(f"\n[v0] Total transactions found: {len(transactions)}")

# トランザクションを日付順にソート
transactions.sort(key=lambda x: x['DateTime (UTC)'])

print("\n[v0] All Transactions:")
print("-" * 80)
for i, tx in enumerate(transactions, 1):
    print(f"\n{i}. Transaction Hash: {tx['Transaction Hash']}")
    print(f"   Status: {tx['Status']}")
    print(f"   Method: {tx['Method']}")
    print(f"   DateTime: {tx['DateTime (UTC)']}")
    print(f"   From: {tx['From']}")
    print(f"   To: {tx['To']}")
    print(f"   Amount: {tx['Amount']}")
    print(f"   Token: {tx['Token']}")

# Mintトランザクションを特定
mint_txs = [tx for tx in transactions if tx['Method'] == 'Mint']
print(f"\n[v0] Mint transactions: {len(mint_txs)}")

# Transferトランザクションを特定
transfer_txs = [tx for tx in transactions if tx['Method'] == 'Transfer']
print(f"[v0] Transfer transactions: {len(transfer_txs)}")

# 送金されたトランザクションを特定（From が 0x0000... でないもの）
outgoing_txs = [tx for tx in transactions if tx['From'] != '0x0000000000000000000000000000000000000000' and tx['From'].lower() != '0x77247cc270768611eb2fbc7759a7b34b9fb045cd'.lower()]
print(f"[v0] Outgoing transactions (funds sent): {len(outgoing_txs)}")

# 受信したトランザクションを特定（To が 0x77247... のもの）
incoming_txs = [tx for tx in transactions if tx['To'].lower() == '0x77247cc270768611eb2fbc7759a7b34b9fb045cd'.lower()]
print(f"[v0] Incoming transactions (funds received): {len(incoming_txs)}")

# 合計金額を計算
total_received = sum(float(tx['Amount'].replace(',', '')) for tx in incoming_txs)
total_sent = sum(float(tx['Amount'].replace(',', '')) for tx in outgoing_txs)

print(f"\n[v0] Total received: {total_received:,.2f} USDC")
print(f"[v0] Total sent: {total_sent:,.2f} USDC")
print(f"[v0] Net balance: {total_received - total_sent:,.2f} USDC")

# test12のキャンペーンに関連するトランザクションを特定
# test12は$300をpledgeして、目標$200を達成しているので、
# クリエイターに$300が送金されているはずです
print("\n[v0] Analysis for test12 campaign:")
print("-" * 80)
print("[v0] Expected: $300 should be transferred to creator after finalize")
print("[v0] Looking for transfer transactions after finalize...")

# 最近のトランザクションを確認
recent_txs = transactions[-5:] if len(transactions) >= 5 else transactions
print(f"\n[v0] Recent transactions (last {len(recent_txs)}):")
for tx in recent_txs:
    print(f"   - {tx['DateTime (UTC)']} | {tx['Method']} | {tx['Amount']} | From: {tx['From'][:10]}... To: {tx['To'][:10]}...")

print("\n[v0] Conclusion:")
if len(outgoing_txs) == 0:
    print("   ⚠️  NO OUTGOING TRANSACTIONS FOUND")
    print("   This means funds have NOT been transferred to the creator.")
    print("   The finalize function may not be releasing funds correctly.")
else:
    print(f"   ✓ Found {len(outgoing_txs)} outgoing transaction(s)")
    print(f"   Total sent: {total_sent:,.2f} USDC")

File: scripts/analyze-transactions.mjs
import fetch from 'node-fetch';

async function analyzeTransactions() {
  try {
    console.log('[v0] Fetching transaction CSV...');
    const response = await fetch('https://hebbkx1anhila5yf.public.blob.vercel-storage.com/export-token-transfer-0x77247CC270768611eb2fBc7759a7b34b9FB045Cd-aNDOvRIizWcT7P9hYKPFvRswZi1dDS.csv');
    
    if (!response.ok) {
      throw new Error(`Failed to fetch CSV: ${response.status} ${response.statusText}`);
    }
    
    const csvText = await response.text();
    console.log('[v0] CSV fetched successfully');
    
    // Parse CSV
    const lines = csvText.split('\n');
    const headers = lines[0].split(',');
    
    console.log('[v0] CSV Headers:', headers);
    console.log('[v0] Total transactions:', lines.length - 1);
    
    // Parse transactions
    const transactions = [];
    for (let i = 1; i < lines.length; i++) {
      if (!lines[i].trim()) continue;
      
      const values = lines[i].split(',');
      const tx = {
        hash: values[0],
        status: values[1],
        method: values[2],
        blockNo: values[3],
        dateTime: values[4],
        from: values[5],
        fromNametag: values[6],
        to: values[7],
        toNametag: values[8],
        amount: values[9],
        valueUSD: values[10],
        token: values[11]
      };
      transactions.push(tx);
    }
    
    console.log('\n[v0] ===== TRANSACTION ANALYSIS =====\n');
    
    // Analyze transactions
    const walletAddress = '0x77247CC270768611eb2fBc7759a7b34b9FB045Cd';
    const mintTxs = transactions.filter(tx => tx.method === 'Mint');
    const transferTxs = transactions.filter(tx => tx.method !== 'Mint');
    const outgoingTxs = transactions.filter(tx => 
      tx.from.toLowerCase() === walletAddress.toLowerCase()
    );
    const incomingTxs = transactions.filter(tx => 
      tx.to.toLowerCase() === walletAddress.toLowerCase()
    );
    
    console.log('[v0] Total Transactions:', transactions.length);
    console.log('[v0] Mint Transactions:', mintTxs.length);
    console.log('[v0] Transfer Transactions:', transferTxs.length);
    console.log('[v0] Outgoing Transactions:', outgoingTxs.length);
    console.log('[v0] Incoming Transactions:', incomingTxs.length);
    
    console.log('\n[v0] ===== MINT TRANSACTIONS =====\n');
    mintTxs.forEach(tx => {
      console.log(`[v0] ${tx.dateTime} | ${tx.amount} | To: ${tx.to}`);
    });
    
    console.log('\n[v0] ===== OUTGOING TRANSACTIONS =====\n');
    outgoingTxs.forEach(tx => {
      console.log(`[v0] ${tx.dateTime} | ${tx.amount} | From: ${tx.from} | To: ${tx.to} | Method: ${tx.method}`);
    });
    
    console.log('\n[v0] ===== INCOMING TRANSACTIONS =====\n');
    incomingTxs.forEach(tx => {
      console.log(`[v0] ${tx.dateTime} | ${tx.amount} | From: ${tx.from} | To: ${tx.to} | Method: ${tx.method}`);
    });
    
    // Calculate totals
    const totalIncoming = incomingTxs.reduce((sum, tx) => {
      const amount = parseFloat(tx.amount.replace(/,/g, ''));
      return sum + (isNaN(amount) ? 0 : amount);
    }, 0);
    
    const totalOutgoing = outgoingTxs.reduce((sum, tx) => {
      const amount = parseFloat(tx.amount.replace(/,/g, ''));
      return sum + (isNaN(amount) ? 0 : amount);
    }, 0);
    
    console.log('\n[v0] ===== SUMMARY =====\n');
    console.log(`[v0] Total Incoming: ${totalIncoming.toLocaleString()} USDC`);
    console.log(`[v0] Total Outgoing: ${totalOutgoing.toLocaleString()} USDC`);
    console.log(`[v0] Net Balance: ${(totalIncoming - totalOutgoing).toLocaleString()} USDC`);
    
    // Check for test12 campaign (300 USDC)
    console.log('\n[v0] ===== TEST12 CAMPAIGN ANALYSIS =====\n');
    const test12Amount = 300;
    const matchingTxs = outgoingTxs.filter(tx => {
      const amount = parseFloat(tx.amount.replace(/,/g, ''));
      return amount === test12Amount;
    });
    
    if (matchingTxs.length > 0) {
      console.log(`[v0] Found ${matchingTxs.length} transaction(s) matching test12 amount (${test12Amount} USDC):`);
      matchingTxs.forEach(tx => {
        console.log(`[v0] ${tx.dateTime} | ${tx.amount} | To: ${tx.to} | Hash: ${tx.hash}`);
      });
    } else {
      console.log(`[v0] ⚠️ NO transactions found matching test12 amount (${test12Amount} USDC)`);
      console.log('[v0] This suggests the funds have NOT been released to the creator.');
    }
    
  } catch (error) {
    console.error('[v0] Error analyzing transactions:', error);
  }
}

analyzeTransactions();
